kv/kvnemesis: TestKVNemesisMultiNode_BufferedWrites failed

kv/kvnemesis.TestKVNemesisMultiNode_BufferedWrites [failed](https://tanzanite.cluster.engflow.com/invocations/default/2ecc89bf-254f-4634-866a-f1c6cbf73be0?testReportRun=1&testReportShard=1&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZuZW1lc2lzOmt2bmVtZXNpc190ZXN0) on master @ [a53b28b8a8d1f1f1544f036bfbdb0952e5371388](https://github.com/cockroachdb/cockroach/commits/a53b28b8a8d1f1f1544f036bfbdb0952e5371388):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x1339c70]

```
Stack: 

```
goroutine 19807 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc0098128d8?, {0x6aa09f8, 0xc0085344e0})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x51b9cc0?, 0xa145d40?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes.func1()
	pkg/kv/kvserver/store_send.go:104 +0x148
panic({0x51b9cc0?, 0xa145d40?})
	GOROOT/src/runtime/panic.go:791 +0x132
github.com/cockroachdb/cockroach/pkg/kv/kvpb.validateExclusionTimestampForBatch({_, _}, {{0x184c3279a95b7a8f, 0x0}, 0x0, {0x184c3279a95be78b, 0x0}, {0x3, 0x3, 0x3, ...}, ...})
	pkg/kv/kvpb/api.go:2572 +0x30
github.com/cockroachdb/cockroach/pkg/kv/kvpb.(*DeleteRequest).Validate(_, {{0x184c3279a95b7a8f, 0x0}, 0x0, {0x184c3279a95be78b, 0x0}, {0x3, 0x3, 0x3, 0x0}, ...})
	pkg/kv/kvpb/api.go:2548 +0x46
github.com/cockroachdb/cockroach/pkg/kv/kvserver/batcheval.Delete({_, _}, {_, _}, {{0x6b41f40, 0xc0016da500}, {{0x184c3279a95b7a8f, 0x0}, 0x0, {0x184c3279a95be78b, ...}, ...}, ...}, ...)
	pkg/kv/kvserver/batcheval/cmd_delete.go:31 +0x125
github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateCommand({_, _}, {_, _}, {_, _}, _, _, {{0x184c3279a95b7a8f, 0x0}, ...}, ...)
	pkg/kv/kvserver/replica_evaluate.go:543 +0x34a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.evaluateBatch({_, _}, {_, _}, {_, _}, {_, _}, _, 0xc003d85680, ...)
	pkg/kv/kvserver/replica_evaluate.go:356 +0xb65
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWrapper(_, {_, _}, {_, _}, {_, _}, _, _, 0xc003f8f2b0, ...)
	pkg/kv/kvserver/replica_write.go:751 +0x1de
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatchWithServersideRefreshes(_, {_, _}, {_, _}, {_, _}, _, _, 0xc003f8f2b0, ...)
	pkg/kv/kvserver/replica_write.go:716 +0x359
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluate1PC(_, {_, _}, {_, _}, _, _, _)
	pkg/kv/kvserver/replica_write.go:564 +0x485
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateWriteBatch(_, {_, _}, {_, _}, _, _, _, {{0x184c3279a2d1e977, 0x0}, ...})
	pkg/kv/kvserver/replica_write.go:443 +0x1b5
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evaluateProposal(0xc004fe4388, {0x6aa0de8, 0xc005798be0}, {0xc002f5c7b0, 0x8}, 0xc003d85540, 0xc003f8f2b0, 0xc0098180b8, {{0x184c3279a2d1e977, 0x0}, ...})
	pkg/kv/kvserver/replica_proposal.go:999 +0x1c9
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).requestToProposal(0xc004fe4388, {0x6aa0de8, 0xc005798be0}, {0xc002f5c7b0, 0x8}, 0xc003d85540, 0xc003f8f2b0, 0xc0098180b8, {{0x184c3279a2d1e977, 0x0}, ...})
	pkg/kv/kvserver/replica_proposal.go:1095 +0x10c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).evalAndPropose(0xc004fe4388, {0x6aa0de8, 0xc005798be0}, 0xc003d85540, 0xc003f8f2b0, 0xc0098180b8, {{0x184c3279a2d1e977, 0x0}, {0x184c3279a23fa747, 0x0}}, ...)
	pkg/kv/kvserver/replica_raft.go:117 +0x156
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeWriteBatch(0xc004fe4388, {0x6aa0de8, 0xc005798a00}, 0xc003d85540, 0xc003f8f2b0)
	pkg/kv/kvserver/replica_write.go:188 +0x6ec
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).executeBatchWithConcurrencyRetries(0xc004fe4388, {0x6aa0de8, 0xc005798a00}, 0xc003d85540, 0x5dbe6a8)
	pkg/kv/kvserver/replica_send.go:510 +0x3a3
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).SendWithWriteBytes(0xc004fe4388, {0x6aa0de8, 0xc005a56c78}, 0xc002ff6840)
	pkg/kv/kvserver/replica_send.go:189 +0x750
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).SendWithWriteBytes(0xc004511408, {0x6aa0de8?, 0xc005a56be0?}, 0xc002ff6840)
	pkg/kv/kvserver/store_send.go:188 +0x83c
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).SendWithWriteBytes(0xc00541b900, {0x6aa0de8, 0xc005a56be0}, 0xc002ff6840)
	pkg/kv/kvserver/stores.go:199 +0x7e
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchInternal(0xc002103508, {0x6aa0de8?, 0xc005a56b48?}, {0xc009819468?}, 0xc002ff6840)
	pkg/server/node.go:1654 +0x55d
github.com/cockroachdb/cockroach/pkg/server.(*Node).Batch(0xc002103508, {0x6aa09f8, 0xc0085344e0}, 0xc002ff6840)
	pkg/server/node.go:1847 +0x286
github.com/cockroachdb/cockroach/pkg/server.(*Node).batchStreamImpl(0xc002103508, {0x7f6aa5a7b8a8, 0xc00bc6a250})
	pkg/server/node.go:1894 +0xae
github.com/cockroachdb/cockroach/pkg/server.(*Node).BatchStream(0xc002103508?, {0x6ae6db0?, 0xc00bc6a250?})
	pkg/server/node.go:1873 +0x3e
github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_BatchStream_Handler({0x59a65a0?, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0})
	bazel-out/k8-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:11494 +0xd8
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.StreamServerInterceptor.func13({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0}, 0xc008850a08, 0x5dbd798)
	pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:143 +0x4ee
google.golang.org/grpc.getChainStreamHandler.func1({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0})
	external/org_golang_google_grpc/server.go:1504 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0}, 0xc008850a08?, 0xc00bc7e140)
	pkg/rpc/context.go:187 +0x6c
google.golang.org/grpc.getChainStreamHandler.func1({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0})
	external/org_golang_google_grpc/server.go:1504 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor({0xc002ba8000?, {{0x449025?}, {0x6af2200?, 0xc002c30180?}}}, {0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0}, 0xc008850a08, 0xc00bc7e100)
	pkg/rpc/auth.go:152 +0x37a
google.golang.org/grpc.getChainStreamHandler.func1({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0})
	external/org_golang_google_grpc/server.go:1504 +0xb2
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0x10?, 0x6aa09f8?})
	pkg/rpc/context.go:145 +0x27
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc0027377a0, {0x6aa09f8, 0xc0085344e0}, {0xc008b9c9c0, 0x27}, 0xc0070cfab0)
	pkg/util/stop/stopper.go:349 +0x13c
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0}, 0xc008850a08, 0xc00bc7e0c0)
	pkg/rpc/context.go:144 +0xcd
google.golang.org/grpc.NewServer.chainStreamServerInterceptors.chainStreamInterceptors.func2({0x59a65a0, 0xc002103508}, {0x6ad2f40, 0xc00bc361e0}, 0xc008850a08, 0xc00bc6a240?)
	external/org_golang_google_grpc/server.go:1495 +0x85
google.golang.org/grpc.(*Server).processStreamingRPC(0xc0051dd860, {0x6af2c20, 0xc007e8b860}, 0xc003fcefc0, 0xc006769140, 0xa21a200, 0x0)
	external/org_golang_google_grpc/server.go:1659 +0x1285
google.golang.org/grpc.(*Server).handleStream(0xc0051dd860, {0x6af2c20, 0xc007e8b860}, 0xc003fcefc0, 0x0)
	external/org_golang_google_grpc/server.go:1739 +0x999
google.golang.org/grpc.(*Server).serveStreams.func1.1()
	external/org_golang_google_grpc/server.go:970 +0xaf
created by google.golang.org/grpc.(*Server).serveStreams.func1 in goroutine 14177
	external/org_golang_google_grpc/server.go:981 +0x136
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestKVNemesisMultiNode_BufferedWrites
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestKVNemesisMultiNode_BufferedWrites968639614
    test_log_scope.go:76: use -show-logs to present logs inline
    kvnemesis_test.go:425: seed: 7881580496811433557
    kvnemesis_test.go:248: kvnemesis logging to /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvnemesis/kvnemesis_test/run_1_of_10/test.outputs/kvnemesis2453866598
```
</p>
</details>

Parameters:
 - <code>attempt=1</code>
 - <code>run=1</code>
 - <code>shard=1</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestKVNemesisMultiNode_BufferedWrites.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-51802

--- a/pkg/kv/kvpb/api.go
+++ b/pkg/kv/kvpb/api.go
@@ -2562,6 +2562,12 @@ func validateExclusionTimestampForBatch(ts hlc.Timestamp, h Header) error {
 		return nil
 	}
 
+	// If we don't have a transaction, we can't know whether
+	// CanForwardReadTimestamp is valid or not.
+	if h.Txn == nil {
+		return nil
+	}
+
 	// Unless the IsoLevel allows per-statement read snapshots,
 	// CanForwardReadTimestamp implies we haven't served a read, so it makes no
 	// sense that ExpectExclusionSince would be set since it is the result of a

--- a/pkg/kv/kvpb/api_test.go
+++ b/pkg/kv/kvpb/api_test.go
@@ -11,6 +11,7 @@ import (
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/kv/kvnemesis/kvnemesisutil"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/isolation"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
@@ -381,6 +382,35 @@ func TestGetValidate(t *testing.T) {
 			KeyLockingStrength:   lock.Exclusive,
 		}
 		require.NoError(t, getReq.Validate(Header{}))
+	})
+}
+
+func TestDeleteValidate(t *testing.T) {
+	t.Run("ExpectExclusionSinceWithCanForwardReadTimestamp", func(t *testing.T) {
+		delReq := &DeleteRequest{
+			ExpectExclusionSince: hlc.Timestamp{WallTime: 1},
+		}
+		require.Error(t, delReq.Validate(Header{
+			CanForwardReadTimestamp: true,
+			Txn:                     &roachpb.Transaction{},
+		}))
+	})
+	t.Run("ExpectExclusionSinceWithCanForwardReadTimestampAtWeakerIsolation", func(t *testing.T) {
+		delReq := &DeleteRequest{
+			ExpectExclusionSince: hlc.Timestamp{WallTime: 1},
+		}
+		txn := &roachpb.Transaction{}
+		txn.IsoLevel = isolation.ReadCommitted
+		require.NoError(t, delReq.Validate(Header{
+			CanForwardReadTimestamp: true,
+			Txn:                     txn,
+		}))
+	})
+	t.Run("ExpectExclusionSinceWithCanForwardReadTimestampButNoTxn", func(t *testing.T) {
+		delReq := &DeleteRequest{
+			ExpectExclusionSince: hlc.Timestamp{WallTime: 1},
+		}
+		require.NoError(t, delReq.Validate(Header{CanForwardReadTimestamp: true}))
 	})
 }
 

